### DISCUSSION TITLE
How do I profile "Loading config file"

### DISCUSSION THREAD

**Discussion by iacore** [lumeland/lume#611] [Q&A] [answered]
Most of the time of building the site (`deno task lume`) seems to be loading `_config.ts`.

How do I profile this?

**Comment by oscarotero** [comment]
The function to detect the config file is this: https://github.com/lumeland/lume/blob/main/core/utils/lume_config.ts#L55

For profiling, I use to change the `lume/` import map to my local clone:

```diff
"imports": {
-    "lume/": "https://deno.land/x/lume@v2.2.0/
+    "lume/": "../lume/
}
```

Why to you think this is a problem? it's expected to load the `_config.ts` file.

**Reply by iacore** [reply]
i suspect the time is spent on loading all the dependencies, not rendering the site.
i want to profile the time it takes to load all the dependencies, which seems to be handled by Deno.

**Reply by oscarotero** [reply, marked as answer]
You can run

```js
const t = performance.now();
console.log(`${t} ms since start!`);
```

to know the time from Deno's start.
But the dependencies are loaded only once, and then cached.

Another way is using Lume events. For example:

```js
site.addEventListener("beforeBuild", () => console.log(`Prepared to start: ${performance.now()}`);
site.addEventListener("afterLoad", () => console.log(`all files are loaded: ${performance.now()}`);
site.addEventListener("afterRender", () => console.log(`all pages are rendered: ${performance.now()}`);
site.addEventListener("beforeSave", () => console.log(`all processors are runned: ${performance.now()}`);
site.addEventListener("afterBuild", () => console.log(`site build: ${performance.now()}`);

```

**Reply by iacore** [reply]
So I was wrong. Rendering pages took a long time.

```
start: 626.560687
prepared to start: 635.202647
all files are loaded: 639.215176
all pages are rendered: 2408.886866
all processors are runned: 3093.362995
...
229 files generated in 2.54 seconds
```

Can I somehow profile this more? To see which page is rendered the slowest.

**Reply by oscarotero** [reply]
The pages are rendered in parallel, so not sure if it's easy to profile it. What template engine are you using? (vto, mdx, njk...)

**Reply by iacore** [reply]
> The pages are rendered in parallel, so not sure if it's easy to profile it. What template engine are you using? (vto, mdx, njk...)

mostly remark with a katex filter, a hljs filter

```ts
import remark from "lume/plugins/remark.ts"
import remarkSmartypants from "https://esm.sh/remark-smartypants@2.1.0"
import remarkMath from 'https://esm.sh/remark-math@6.0.0'

site.use(remark({
  remarkPlugins: [
    [remarkMath],
    [remarkSmartypants, {}],
    // [remarkObsidianLink, {
    //   hrefTemplate: (link_text: string) => `../${link_text}`
    // }]
  ]
}))
```

**Reply by oscarotero** [reply]
Can you try if with markdown-it goes faster?

**Reply by iacore** [reply]
I switched to markdown-it. Markdown-it is faster.

This time I used hyperfine to measure `deno task build` instead of eyeballing:

**What: (mean ± σ)**
Before (remarked + katex, djot): 3.661 s ±  0.126 s
After (no katex) (markdown-it, djot): 2.852 s ±  0.124 s
After (feature parity) (markdown-it, djot + katex): 3.017 s ±  0.191 s
After (but the document with most katex removed): 2.916 s ±  0.185 s

I think katex is really slow.

[I documented this on my website.](https://www.1a-insec.net/about/site-improvement-history/) Perhaps you want to be attributed to by your name or Github handle or not at all?

**Reply by oscarotero** [reply]
Hi, thanks for sharing it, really interesting info.
- markdown-it is the fastest markdown library for JavaScript (I remember to test other alternatives and all of them were slower).
- as you may know, Lume [has a plugin for katex](https://lume.land/plugins/katex/). I don't know if it's faster than remark/markdown-it plugins (probably all of them use the same katex library under the hood).
- In your changes to decrease the build time, I see you removed the inline and esbuild plugins. Did you have many inlined elements in your site? I'm curious to know if the performance depends on the amount of elements to inline, or only to have the plugin enabled.

Many plugins use DOM API to read and manipulate the HTML (inline plugin is one of them). In Lume I'm using [deno-dom](https://github.com/b-fuze/deno-dom) to do that, but I'm also considering to switch to [LinkeDOM](https://github.com/WebReflection/linkedom) which is faster and consume less memory. It's not possible to do this change now, due some differences that would break some sites, but once these issues are addressed, I think this change would make Lume faster.

**Reply by oscarotero** [reply]
> Perhaps you want to be attributed to by your name or Github handle or not at all?

It's okay as is, but I leave this decision to you :)

**Reply by iacore** [reply]
> * as you may know, Lume [has a plugin for katex](https://lume.land/plugins/katex/). I don't know if it's faster than remark/markdown-it plugins (probably all of them use the same katex library under the hood).

I use a fork of this plugin to handle Djot.js's output.
 
> * In your changes to decrease the build time, I see you removed the inline and esbuild plugins. Did you have many inlined elements in your site? I'm curious to know if the performance depends on the amount of elements to inline, or only to have the plugin enabled.

I have 0 inlined elements.

esbuild does slow the site down. I hope there is caching involved, so I don't have to worry about wasting time on this.

> I'm also considering to switch to [LinkeDOM](https://github.com/WebReflection/linkedom) which is faster and consume less memory.

This will potentially make the build time much much faster. I used [uhtml](https://github.com/WebReflection/uhtml)/dom from the same author, and it can only be described as instant. I modified uhtml/dom to be used with Paper.js to generate SVG files. [The code is here.](https://git.envs.net/iacore/paper.js-deno/)

I think you can easily modify the official plugins to use a light variant of uhtml/dom or linkedom. But make it another API. 

Is it possible to make the current `page.document` a calculated property? If so, migration can be easier. mmm you might end up needing to break the current API in order to make DOM-in-JS go faster.

**Reply by oscarotero** [reply]
@gilesdring has worked on testing LinkeDOM in this branch https://github.com/dringtech/lume/tree/feat/linkedom

You can read the [conversation in Discord](https://discord.com/channels/794537085641818124/1133364035249774634/1227277820082782288).

The change from deno-dom to LinkeDOM it's easy (both packages are based on web standards so they have the same behavior). The only problem [is this issue](https://github.com/WebReflection/linkedom/issues/231). Once it's fixed, I think Lume could switch to LinkeDOM without breaking changes.

**Reply by iacore** [reply]
> You can read the [conversation in Discord](https://discord.com/channels/794537085641818124/1133364035249774634/1227277820082782288).

How do I join said Discord channel?

> The only problem https://github.com/WebReflection/linkedom/issues/231.

Where exactly is `DocumentFragment.textContent` used? Either I can fix the usage or I ~~can~~ [already fixed linkedom](https://github.com/WebReflection/linkedom/pull/278).

**Reply by oscarotero** [reply]
> How do I join said Discord channel?

You can join to Lume server here: https://discord.gg/YbTmpACHWB

> Where exactly is DocumentFragment.textContent used?

You have an example here: https://github.com/lumeland/lume/blob/a42088c694225166b4fb894d1b2c4cef6da407ed/plugins/inline.ts#L47-L63 (probably there are more). The `inline` plugin should work for elements inside and ouside `<template>`  elements.

**Reply by oscarotero** [reply]
hmmm, that's odd. It's infinite:
<img width="727" alt="image" src="https://github.com/lumeland/lume/assets/377873/af392fcd-e30d-447d-955a-f2dd5a307dcc">

Can you try this?
https://discord.gg/wFJvXBdvny
